src/boot/fwupdate.c

```diff
@@ -0,0 +1,595 @@
+
+                *ret_mode = mode = proto->Mode->Mode;
+                found = 1;
+                return EFI_SUCCESS;
+        }
+
+        found = 0;
+        return EFI_UNSUPPORTED;
+}
+
+static EFI_STATUS device_path_find_media_device_path_and_truncate(EFI_DEVICE_PATH *dp, size_t *ret_index) {
+        size_t index = 0;
+
+        assert(ret_index);
+
+        if (!dp) {
+                *ret_index = 0;
+                return EFI_SUCCESS;
+        }
+
+        for (;;) {
+                if (device_path_is_end(dp))
+                        return log_error_status(EFI_INVALID_PARAMETER, "Device path for is empty");
+                if (dp->Type == MEDIA_DEVICE_PATH && dp->SubType == MEDIA_FILEPATH_DP)
+                        break;
+
+                index++;
+                dp = device_path_next_node(dp);
+        }
+
+        *dp = DEVICE_PATH_END_NODE;
+        *ret_index = index;
+
+        return EFI_SUCCESS;
+}
+
+/* Take a full blown devicepath as input, and return a handle to the device this refers to, plus the
+ * remaining suffix of the device path, that is the file path part of it. */
+static EFI_STATUS find_capsule(
+                const char16_t *name,              /* efi variable name this update was configured in */
+                const EFI_DEVICE_PATH *update_dp,  /* full blown device path to look for */
+                EFI_HANDLE *ret_device_handle,     /* handle to device the path was found on */
+                EFI_DEVICE_PATH **ret_suffix_dp) { /* pointer into 'update_dp' pointing to the file system path suffix of the device path */
+
+        EFI_DEVICE_PATH *file_dp = (EFI_DEVICE_PATH *) ASSERT_PTR(update_dp);
+        EFI_HANDLE device_handle;
+        EFI_STATUS err;
+
+        assert(name);
+        assert(ret_device_handle);
+        assert(ret_suffix_dp);
+
+        err = BS->LocateDevicePath(MAKE_GUID_PTR(EFI_SIMPLE_FILE_SYSTEM_PROTOCOL), &file_dp, &device_handle);
+        if (err == EFI_SUCCESS)
+                if ((file_dp)->Type != MEDIA_DEVICE_PATH || (file_dp)->SubType != MEDIA_FILEPATH_DP) {
+                        return log_error_status(EFI_INVALID_PARAMETER, "Invalid device path type for '%ls': %m", name);
+
+                *ret_device_handle = device_handle;
+                return EFI_SUCCESS;
+        }
+
+        _cleanup_free_ EFI_HANDLE *handles = NULL;
+        size_t n_handles;
+        err = BS->LocateHandleBuffer(
+                        ByProtocol,
+                        MAKE_GUID_PTR(EFI_SIMPLE_FILE_SYSTEM_PROTOCOL),
+                        /* SearchKey= */ NULL,
+                        &n_handles,
+                        &handles);
+        if (err != EFI_SUCCESS)
+                return log_error_status(err, "Failed to locate file system handles: %m");
+
+        _cleanup_free_ EFI_DEVICE_PATH *dp = device_path_dup(file_dp);
+        size_t index = 0;
+
+        err = device_path_find_media_device_path_and_truncate(dp, &index);
```

Comment:
i still think this is a really weird thing, and should be reworked as discussed before, and return a duplicated path itself as one ret param, and a pointer into the original dp for the remainder.